### PR TITLE
Add getRoomMembers

### PR DESCRIPTION
**Pull request by Half-Shot**
for calling /members

### CODE REVIEW

File: src/MatrixClient.ts [hunk 1]
```diff
@@ -667,6 +668,7 @@ export class MatrixClient extends EventEmitter {
 
     /**
      * Gets the joined members in a room. The client must be in the room to make this request.
+     * This request is usually faster than getRoomMembers.
```
Comment (turt2live) [inline src/MatrixClient.ts]: This isn't relevant to the API structure and describes an implementation detail - please remove it.

File: src/MatrixClient.ts [hunk 2]
```diff
@@ -677,6 +679,23 @@ export class MatrixClient extends EventEmitter {
         });
     }
 
+        /**
```
Comment (turt2live) [inline src/MatrixClient.ts]: indentation

File: src/MatrixClient.ts [hunk 3]
```diff
@@ -677,6 +679,23 @@ export class MatrixClient extends EventEmitter {
...
+     * Gets the joined members in a room. The client must be in the room to make this request.
```
Comment (turt2live) [inline src/MatrixClient.ts]: It gets any membership, not just joined. Also, the client having to be in the room is implicit or potentially irrelevant in the future.
```suggestion
     * Gets the members in a room. 
```

File: src/MatrixClient.ts [hunk 4]
```diff
@@ -677,6 +679,23 @@ export class MatrixClient extends EventEmitter {
...
+     * @param {string} roomId The room ID to get the joined members of.
```
Comment (turt2live) [inline src/MatrixClient.ts]:
```suggestion
     * @param {string} roomId The room ID to get the members of.
```

File: src/MatrixClient.ts [hunk 5]
```diff
@@ -677,6 +679,23 @@ export class MatrixClient extends EventEmitter {
...
+     * @param {string} membership Get members who are of this type. Leave undefined if no filtering is needed.
+     * @param {string} notMembership Do not get members who are of this type. Leave undefined if no filtering is needed.
+     * @returns {Promise<string>} The joined user IDs in the room
```
Comment (turt2live) [inline src/MatrixClient.ts]:
```suggestion
     * @returns {Promise<string>} The member user IDs in the room
```

File: src/MatrixClient.ts [hunk 6]
```diff
@@ -677,6 +679,23 @@ export class MatrixClient extends EventEmitter {
...
+     */
+    @timedMatrixClientFunctionCall()
+    public getRoomMembers(roomId: string, membership?: MembershipEnum, notMembership?: MembershipEnum) {
```
Comment (turt2live) [inline src/MatrixClient.ts]: This needs a typed return value of `: Promise<string>`

File: src/MatrixClient.ts [hunk 7]
```diff
@@ -677,6 +679,23 @@ export class MatrixClient extends EventEmitter {
...
+        return this.doRequest("GET", "/_matrix/client/r0/rooms/" + encodeURIComponent(roomId) + "/members", {
+            membership,
+            not_membership: notMembership,
+        }).then(response => {
+            return response['chunk'] as MatrixRoomMemberEvent[];
```
Comment (turt2live) [inline src/MatrixClient.ts]: is it a string or a structure? If it's a structure, I have quite a few concerns.

File: src/models/Events.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
+export interface MatrixRoomEvent {
```
Comment (turt2live) [inline src/models/Events.ts]: Nothing in this file is how I want to define events in this SDK - please leave this as a task for https://github.com/turt2live/matrix-js-bot-sdk/issues/11 and similar issues.

File: test/MatrixClientTest.ts [hunk 9]
```diff
@@ -2204,6 +2205,47 @@ describe('MatrixClient', () => {
...
+                      "age": 1234
+                    },
+                    "state_key": "@alice:example.org"
+                  }
+            ];
+
+            http.when("GET", "/_matrix/client/r0/rooms").respond(200, (path, content, req) => {
+                expect(req.opts.qs.membership).toEqual("join");
+                expect(req.opts.qs.not_membership).toEqual("invite");
+                expect(path).toEqual(
```
Comment (turt2live) [inline test/MatrixClientTest.ts]: always check the path first.

File: test/MatrixClientTest.ts [hunk 10]
```diff
@@ -2204,6 +2205,47 @@ describe('MatrixClient', () => {
...
+                    `${hsUrl}/_matrix/client/r0/rooms/${encodeURIComponent(roomId)}/members`
```
Comment (turt2live) [inline test/MatrixClientTest.ts]: this linefeed is weird - let's not do that.

File: test/MatrixClientTest.ts [hunk 11]
```diff
@@ -2204,6 +2205,47 @@ describe('MatrixClient', () => {
...
+                );
+                return {chunk: members};
+            });
+
+            http.flushAllExpected();
+            const result = await client.getRoomMembers(roomId, "join", "invite");
+            expectArrayEquals(members, result);
+        });
+    });
```
Comment (turt2live) [inline test/MatrixClientTest.ts]: there also needs to be tests for when both arguments are undefined and when either is undefined.

File: src/models/Events.ts (hunk 8 above)
Comment (Half-Shot) [reply src/models/Events.ts]: It would be much easier if you actually defined how you'd want them to look.

**Comment by Half-Shot** [comment]
I'll probably leave this as blocked until some form of typing emerges, as /members returns a sufficiently complex type that it would benefit from a typed system.

File: src/models/Events.ts (hunk 8 above)
Comment (turt2live) [reply src/models/Events.ts]: correct, but I don't really know what that is yet ._.

Just "not this" unfortunately.

**Comment by turt2live** [comment]
Thank you for doing this @Half-Shot, though I've implemented this as https://github.com/turt2live/matrix-js-bot-sdk/commit/753ce2c33b05699f24925f241fd4ee39db954211 and https://github.com/turt2live/matrix-js-bot-sdk/commit/dd3d580c2e38f03b9b707fe78a9058afa186cc11 - which includes the type definitions I was thinking for #11 (kinda-mostly).

**Pull request closed without merging**